proof/refine/RISCV64/KHeap_R.thy

```diff
@@ -4307,6 +4364,178 @@ lemma update_sched_context_rewrite:
+                    simp: snd_bind)
+  done
+
+lemma getSchedContext_setSchedContext_decompose:
+   "monadic_rewrite False True
+     (sc_at' scPtr and K (\<forall>sc. objBits (f sc) = objBits sc) and K (\<forall>sc. objBits (g sc) = objBits sc))
+     (do sc \<leftarrow> getSchedContext scPtr;
+         setSchedContext scPtr (g (f sc))
+      od)
+     (do sc \<leftarrow> getSchedContext scPtr;
+         setSchedContext scPtr (f sc);
+         sc \<leftarrow> getSchedContext scPtr;
+         setSchedContext scPtr (g sc)
+      od)"
+  apply (clarsimp simp: monadic_rewrite_def)
+  apply (rule monad_state_eqI)
+    apply (simp add: in_monad getSchedContext_def getObject_def)
+    apply (frule no_ofailD[OF no_ofail_sc_at'_readObject])
+    apply (clarsimp del: readObject_misc_ko_at' simp del: readObject_misc_obj_at')
+    apply (clarsimp simp: setSchedContext_def setObject_def obj_at'_def projectKOs objBits_simps'
+                          in_monad RISCV64_H.fromPPtr_def scBits_pos_power2 updateObject_default_def
+                          in_magnitude_check ps_clear_upd magnitudeCheck_assert split_def
+                     del: readObject_misc_ko_at'
+                   split: option.split_asm)
+     apply (rename_tac sc sc')
+     apply (rule_tac x="f sc" in exI)
+     apply (rule conjI;
+            fastforce simp: readObject_def obind_def omonad_defs split_def RISCV64_H.fromPPtr_def
+                            ps_clear_upd loadObject_default_def lookupAround2_known1 projectKOs
+                            objBits_simps' scBits_pos_power2 lookupAround2_None2 lookupAround2_char2
+                     split: option.splits if_split_asm dest!: readObject_misc_ko_at')
+    apply (rename_tac sc p sc')
+    apply (rule_tac x="f sc" in exI)
+    apply (rule conjI)
+     apply (thin_tac "scBitsFromRefillLength' _=_")
+     apply (clarsimp simp: readObject_def obind_def omonad_defs fun_upd_def split_def RISCV64_H.fromPPtr_def
+                           ps_clear_upd loadObject_default_def lookupAround2_known1 projectKOs
+                           objBits_simps' scBits_pos_power2 lookupAround2_None2 lookupAround2_char2
+                    split: option.splits if_split_asm)
+     apply (metis option.simps(3) word_le_less_eq word_le_not_less)
+    apply (clarsimp simp: split: option.splits)
+    apply (metis (no_types) array_rules(2) lookupAround2_char2 mcs(1) order.strict_trans2
+                            word_le_less_eq word_le_not_less)
+   apply (simp add: in_monad getSchedContext_def getObject_def)
+   apply (frule no_ofailD[OF no_ofail_sc_at'_readObject])
+   apply (clarsimp del: readObject_misc_ko_at' simp del: readObject_misc_obj_at')
+   apply (clarsimp simp: setSchedContext_def setObject_def projectKOs in_monad ps_clear_upd obj_at'_def
+                         split_def updateObject_default_def magnitudeCheck_assert RISCV64_H.fromPPtr_def
+                  dest!: readObject_misc_ko_at')
+
```

Comment:
I don't know if the extra empty lines help in this lemma, might be worth dropping them.

Reply:
I'm a big fan of extra lines in nasty proofs like this one. The indentation does already say "yes this is the end of that subgoal", of course, but I do like the extra separation. Any further thoughts on what to do with extra lines?

Reply:
Fine to leave them if they add something. Looking over this specific proof, I didn't see the benefit, but you'd know better than me.